opt_einsum/sharing.py

```diff
@@ -4,10 +4,37 @@
 import functools
 import numbers
 from collections import Counter
+import threading
 
 from .parser import alpha_canonicalize, parse_einsum_input
 
-_SHARING_STACK = []
+
+_SHARING_STACK = {}
+
+
+def currently_sharing():
+    """Check if we are currently sharing a cache -- thread specific.
+    """
+    return threading.get_ident() in _SHARING_STACK
+
+
+def get_sharing_cache():
+    """Return the most recent sharing cache -- thread specific.
+    """
+    return _SHARING_STACK[threading.get_ident()][-1]
+
+
+def _add_sharing_cache(cache):
+    tid = threading.get_ident()
+    _SHARING_STACK.setdefault(tid, [])
+    _SHARING_STACK[tid].append(cache)
+
+
+def _remove_sharing_cache(cache):
+    tid = threading.get_ident()
+    _SHARING_STACK[tid].remove(cache)
```

Comment:
What is the advantage of `.remove()` over `.pop()`? If these are really stacks, then we should use `.pop()`. If these are not really stacks, then we should use sets or something.

Reply:
Yes should be exactly the same for the current behaviour, ``remove(cache)`` felt more explicit but is probably unnecessary, I'll update it.